File: examples/openapi-sdk-petstore/kotlin/sdk/com/example/petstore/api/RequestOptions.kt
package com.example.petstore.api

import okhttp3.HttpUrl
import okhttp3.Request
import okhttp3.Response

/**
 * Per-call options bag — every generated method takes a
 * `RequestOptions` last param. Mirrors hey-api's TS SDK options shape.
 *
 *  - `client`               — override the impl's bound `APIClient`.
 *  - `baseUrl`              — override `client.baseUrl` for one call.
 *  - `timeout`              — per-call call-timeout in milliseconds.
 *  - `headers`              — extra/override headers, applied last.
 *  - `requestInterceptors`  — extra suspend interceptors, run after
 *                              client-level ones.
 *  - `responseValidator`    — runs after a 2xx response; throwing
 *                              converts the call into an exception.
 *  - `responseTransformer`  — rewrite the response body before decode.
 */
public data class RequestOptions(
    public val client: APIClient? = null,
    public val baseUrl: HttpUrl? = null,
    public val timeout: Long? = null,
    public val headers: Map<String, String> = emptyMap(),
    public val requestInterceptors: List<suspend (Request) -> Request> = emptyList(),
    public val responseValidator: (suspend (ByteArray, Response) -> Unit)? = null,
    public val responseTransformer: (suspend (ByteArray) -> ByteArray)? = null,
)

File: examples/openapi-sdk-petstore/go/sdk/request_options.go
package petstore

import (
	"context"
	"net/http"
	"time"
)

// RequestOptions is the per-call options bag every generated method
// takes as its trailing argument. Mirrors hey-api's TS SDK options
// shape.
//
//   - Client overrides the impl's bound *APIClient for one call.
//   - BaseURL overrides client.BaseURL for one call.
//   - Timeout, when non-zero, derives a context with that deadline
//     from the call's ctx. Use ctx.WithDeadline yourself for finer
//     control; this is the convenience knob.
//   - Headers are extra/override headers, applied last (so callers
//     can override Content-Type / auth headers if they want).
//   - RequestInterceptors run after client-level ones.
//   - ResponseValidator runs after a 2xx response; returning an error
//     converts the call into a failure.
//   - ResponseTransformer rewrites the response body before decoding.
type RequestOptions struct {
	Client               *APIClient
	BaseURL              string
	Timeout              time.Duration
	Headers              map[string]string
	RequestInterceptors  []func(*http.Request) (*http.Request, error)
	ResponseValidator    func(body []byte, resp *http.Response) error
	ResponseTransformer  func(body []byte) ([]byte, error)
}

// applyTimeout returns a context derived from ctx with opts.Timeout
// as a deadline. Returns ctx unchanged when Timeout is zero.
func (opts RequestOptions) applyTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
	if opts.Timeout <= 0 {
		return ctx, func() {}
	}
	return context.WithTimeout(ctx, opts.Timeout)
}
